### PR TITLE
kv: add cache to snapshot.

### PR DESCRIPTION
**Pull request by disksing**
1. Add CacheSnapshot. CacheSnapshot wraps a snapshot and provides some cache mechanisms.
2. Add batch operations to `Transaction` for future performance optimizations.

### CODE REVIEW

**Comment by disksing** [comment]
@c4pt0r @ngaut @siddontang PTAL.
This PR is expected to be merged into hbase branch, so please ignore CI build message (due to dependency problems).I have passed all tests locally.

**Comment by ngaut** [comment]
You need to pass CI first.

File: store/hbase/snapshot.go [hunk 1]
```diff
@@ -72,6 +72,50 @@ func (s *hbaseSnapshot) Get(k kv.Key) ([]byte, error) {
...
+	scanner := s.txn.GetScanner([]byte(s.storeName), start, end, limit)
+	defer scanner.Close()
+
+	m := make(map[string][]byte)
+	for i := 0; i < limit; i++ {
+		r := scanner.Next()
+		if r != nil && len(r.Columns) > 0 {
+			k := r.Row
+			v := r.Columns[FmlAndQual].Value
+			m[string(k)] = v
```
Comment (ngaut) [inline store/hbase/snapshot.go]: Do not create string twice.

File: kv/kv.go [hunk 2]
```diff
@@ -101,6 +101,10 @@ var ErrNotCommitted = errors.New("this transaction is not committed")
 type Transaction interface {
 	// Get gets the value for key k from KV store.
 	Get(k Key) ([]byte, error)
+	// BatchGet gets a batch of values from KV store.
+	BatchGet(keys []Key) (map[string][]byte, error)
+	// Scan gets values in specific range from KV store.
+	Scan(start, end Key, limit int) (map[string][]byte, error)
```
Comment (c4pt0r) [inline kv/kv.go]: If your return value doesn't have ordered semantic, do not use 'Scan' for function name.

File: kv/union_store.go [hunk 3]
```diff
@@ -52,34 +34,26 @@ func IsErrNotFound(err error) bool {
...
-	// Set associates key with value
-	Set([]byte, []byte) error
-}
-
-// UnionStore is an implement of Store which contains a buffer for update.
+// UnionStore is an in-memory Store which contains a buffer for write and a
+// snapshot for read.
 type UnionStore struct {
-	Dirty    MemBuffer // updates are buffered in memory
+	WBuffer  MemBuffer // updates are buffered in memory
```
Comment (c4pt0r) [inline kv/union_store.go]: WBuffer means WriteBuffer?

Comment (disksing) [inline kv/union_store.go]: Yes. This name was suggested by @ngaut .

**Comment by c4pt0r** [comment]
Leave minor comments, the rest LGTM.

File: kv/cache_snapshot_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	c.Assert(bytes.Compare(v, s.v4), Equals, 0)
+}
+
+func (s *testCacheSnapshotSuite) TestBatchGet(c *C) {
+	s.store.Set(s.k1, s.v1)
+	s.store.Set(s.k2, s.v2)
+
+	m, err := s.cache.BatchGet([]Key{s.k1, s.k2, s.k3})
+	c.Assert(err, IsNil)
+	c.Assert(bytes.Compare(m[string(s.k1)], s.v1), Equals, 0)
```
Comment (siddontang) [inline kv/cache_snapshot_test.go]: use DeepEqual or ByteEquals for bytes equal comparison, the same for all followings.

File: kv/kv.go (hunk 2 above)
Comment (siddontang) [inline kv/kv.go]: what's the scan range scope? [start, end] or [start, end), you should point it out to let user know.

Comment (disksing) [inline kv/kv.go]: OK, I will switch to `RangeGet()`.

File: kv/union_store_test.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(bytes.Compare(v, s.v2), Equals, 0)
+}
+
+func (s *testUnionStoreSuite) TestDelete(c *C) {
+	s.store.Set(s.k1, s.v1)
+	err := s.us.Delete(s.k1)
+	c.Assert(err, IsNil)
+	_, err = s.us.Get(s.k1)
+	c.Assert(IsErrNotFound(err), IsTrue)
```
Comment (siddontang) [inline kv/union_store_test.go]: how about supporting ErrorEquals in check for error check? @qiuyesuifeng

File: store/hbase/snapshot.go [hunk 6]
```diff
@@ -72,6 +72,50 @@ func (s *hbaseSnapshot) Get(k kv.Key) ([]byte, error) {
 	return nil, kv.ErrNotExist
 }
 
+// BatchGet implements kv.Snapshot.BatchGet().
+func (s *hbaseSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
+	gets := make([]*hbase.Get, len(keys))
+	for i, key := range keys {
+		g := hbase.NewGet(key)
+		g.AddColumn([]byte(ColFamily), []byte(Qualifier))
```
Comment (siddontang) [inline store/hbase/snapshot.go]: you should define a slice variable, no need to convert the const string every time.

**Comment by disksing** [comment]
PTAL @c4pt0r @siddontang @ngaut

File: kv/cache_snapshot.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	for k, v := range values {
+		c.cache.Set([]byte(k), v)
+		m[k] = v
+	}
+	return m, nil
+}
+
+// RangeGet gets values from snapshot and saves them in cache.
+func (c *cacheSnapshot) RangeGet(start, end Key, limit int) (map[string][]byte, error) {
```
Comment (ngaut) [inline kv/cache_snapshot.go]: Is it [start, end] or [start, end) ? We should comment it.

File: kv/cache_snapshot_test.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+func (s *testCacheSnapshotSuite) TestRangeGet(c *C) {
+	s.store.Set(s.k1, s.v1)
+	s.store.Set(s.k2, s.v2)
+	s.store.Set(s.k3, s.v3)
+
+	m, err := s.cache.RangeGet(s.k1, s.k2, 100)
+	c.Assert(err, IsNil)
+	c.Assert(len(m), Equals, 2)
```
Comment (siddontang) [inline kv/cache_snapshot_test.go]: Use HasLen checker instead

File: kv/cache_snapshot_test.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+import (
+	"github.com/juju/errors"
+	. "github.com/pingcap/check"
+)
+
+var _ = Suite(&testCacheSnapshotSuite{})
+
+type testCacheSnapshotSuite struct {
+	k1, k2, k3, k4 []byte
```
Comment (ngaut) [inline kv/cache_snapshot_test.go]: Why don't you just pair key and value ?

Comment (disksing) [inline kv/cache_snapshot_test.go]: Keys and values are not always in pairs during the test. These variables are just used for identifying different values.

File: kv/union_store_test.go (hunk 5 above)
Comment (disksing) [inline kv/union_store_test.go]: I think we may implement a `Checker` in `terror` package. Then we are able to do something like `c.Assert(err, terror.Equal, kv.ErrNotExist)`. @siddontang @qiuyesuifeng

File: store/hbase/txn.go [hunk 10]
```diff
@@ -103,6 +103,18 @@ func (txn *hbaseTxn) Get(k kv.Key) ([]byte, error) {
...
 }
 
+func (txn *hbaseTxn) BatchGet(keys []kv.Key) (map[string][]byte, error) {
+	encodedKeys := make([]kv.Key, len(keys))
+	for i, k := range keys {
+		encodedKeys[i] = kv.EncodeKey(k)
+	}
+	return txn.UnionStore.Snapshot.BatchGet(encodedKeys)
+}
+
```
Comment (coocood) [inline store/hbase/txn.go]: I suppose we should get from WBuffer first, then gets those missing keys in Snapshot.

File: util/batch_worker.go [hunk 11]
```diff
@@ -20,6 +20,7 @@ import (
 
 type batchWorkFunc func(jobs []interface{})
 
+// BatchWorker collects jobs then batch process them.
```
Comment (siddontang) [inline util/batch_worker.go]: and them processes them in batches.

Comment (disksing) [inline util/batch_worker.go]: This utility is actually useless. It will be removed in my next PR.

File: store/hbase/txn.go (hunk 10 above)
Comment (disksing) [inline store/hbase/txn.go]: Yes, we may get wrong values here. As a caller of Transaction actually is not really care about specific values, I will change function names of Transaction from `BatchGet` to `BatchPrefetch`.

**Comment by disksing** [comment]
PTAL @ngaut @coocood @c4pt0r @siddontang

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**